libcontainer/container_linux.go

```diff
@@ -500,117 +615,181 @@ func (c *Container) commandTemplate(p *Process, childInitPipe *os.File, childLog
+			// from a root filesystem with config.RootPropagation applied. This
+			// is needed because existing runc users expect bind-mounts to act
+			// this way (the "classic" method), regardless of the mount API
+			// used to create the bind-mount.
+			//
+			// NOTE: This explicitly does not handle configurations where there
+			// is a bind-mount source of a path from inside the container
+			// rootfs. We could do this in rootfs_linux.go but this would
+			// require doing criu-style shennanigans to try to figure out how
+			// to recreate the state in /proc/self/mountinfo. For now, it's
+			// much simpler to implement this based purely on RootPropagation.
+
+			// Same logic as prepareRoot().
+			propFlags := unix.MS_SLAVE | unix.MS_REC
+			if c.config.RootPropagation != 0 {
+				propFlags = c.config.RootPropagation
+			}
 
-	return c.sendFdsSources(cmd, messageSockPair, "_LIBCONTAINER_IDMAP_FDS", func(m *configs.Mount) bool {
-		return m.IsBind() && m.IsIDMapped()
-	})
-}
+			// The one thing to consider is whether the RootPropagation is
+			// MS_REC and whether the mount source is the same as /. If it is
+			// MS_REC then we apply the propagation flags (nix MS_REC)
+			// recursively. Otherwise we apply the propagation flags
+			// non-recursively if m.Source is part of the / mount. We do
+			// nothing if neither is the case.
+			var setattrFlags uint
+			if propFlags&unix.MS_REC == unix.MS_REC {
+				// If the RootPropagation is recursive, then any bind-mount
+				// from the host should inherit the propagation setting of /.
+				logrus.Debugf("remap mount sources: applying recursive RootPropagation of 0x%x to %q bind-mount propagation", propFlags, m.Source)
+				setattrFlags = unix.AT_RECURSIVE
+			} else {
+				// If the RootPropagation is not recursive, only bind-mounts
+				// from paths within the / mount should inherit the propagation
+				// setting /.
+				info, err := mountinfo.GetMounts(mountinfo.ParentsFilter(m.Source))
+				if err != nil {
+					return fmt.Errorf("remap mount sources: failed to parse /proc/self/mountinfo: %w", err)
+				}
+				// If there is only a single entry with a mountpoint path of /,
+				// the source was a child of /. Otherwise, do not set the
+				// propagation setting of bind-mounts.
+				if len(info) == 1 && info[0].Mountpoint == "/" {
+					logrus.Debugf("remap mount sources: applying non-recursive RootPropagation of 0x%x to child of / %q bind-mount propagation", propFlags, m.Source)
+				} else {
+					logrus.Debugf("remap mount sources: using default mount propagation for %q bind-mount", m.Source)
+					propFlags = 0
+				}
+			}
+			if propFlags != 0 {
+				if err := unix.MountSetattr(int(mountFile.Fd()), "", unix.AT_EMPTY_PATH|setattrFlags, &unix.MountAttr{
+					Propagation: uint64(propFlags &^ unix.MS_REC),
+				}); err != nil {
+					return fmt.Errorf("remap mount sources: failed to set mount propagation of %q bind-mount to 0x%x: %w", m.Source, propFlags, err)
```

Comment:
It will be nice to say which syscall is failing, as the error can be not implemented and it is not obvious without grepping the code which syscall is missing